File: AlgoExpert/Easy/Sorted_Squared_Array/CPP/Sorted_Squared_Array.cpp
/*
Problem Name: Sorted Squared Array
Difficulty: Easy
Platform: AlgoExpert

Problem Description:
Write a function that takes in a non-empty array of integers that are sorted in ascending order
and returns a new array of the same length with the squares of the original integers also sorted in ascending order.

Time Complexity: O(n)
Space Complexity: O(n)
*/

#include <vector>
#include <cmath>
using namespace std;

vector<int> sortedSquaredArray(vector<int>& array) {
    int n = array.size();
    vector<int> sortedSquares(n, 0);
    int left = 0, right = n - 1;

    for (int i = n - 1; i >= 0; --i) {
        int leftValue = array[left];
        int rightValue = array[right];
        if (abs(leftValue) > abs(rightValue)) {
            sortedSquares[i] = leftValue * leftValue;
            left++;
        } else {
            sortedSquares[i] = rightValue * rightValue;
            right--;
        }
    }
    return sortedSquares;
}

File: AlgoExpert/Easy/Sorted_Squared_Array/Python/Sorted_Squared_Array.py
"""
Problem Name: Sorted Squared Array
Difficulty: Easy
Platform: AlgoExpert

Problem Description:
Write a function that takes in a non-empty array of integers that are sorted in ascending order
and returns a new array of the same length with the squares of the original integers also sorted in ascending order.

Time Complexity: O(n)
Space Complexity: O(n)
"""

def sortedSquaredArray(array):
    sortedSquares = [0 for _ in array]
    leftIdx = 0
    rightIdx = len(array) - 1

    for i in reversed(range(len(array))):
        leftValue = array[leftIdx]
        rightValue = array[rightIdx]
        if abs(leftValue) > abs(rightValue):
            sortedSquares[i] = leftValue * leftValue
            leftIdx += 1
        else:
            sortedSquares[i] = rightValue * rightValue
            rightIdx -= 1

    return sortedSquares

# Example usage
if __name__ == "__main__":
    array = [1, 2, 3, 5, 6, 8, 9]
    result = sortedSquaredArray(array)
    print("Sorted squared array:", result)
